core/state_transition.go

```diff
@@ -252,3 +262,105 @@ func (st *StateTransition) refundGas() {
 func (st *StateTransition) gasUsed() uint64 {
 	return st.initialGas - st.gas
 }
+
+// StakingTransitionDb will transition the state by applying the staking message and
+// returning the result including the used gas. It returns an error if failed.
+// It is used for staking transaction only
+func (st *StateTransition) StakingTransitionDb() (usedGas uint64, err error) {
+	if err = st.preCheck(); err != nil {
+		return 0, err
+	}
+	msg := st.msg
+	sender := vm.AccountRef(msg.From())
+	homestead := st.evm.ChainConfig().IsS3(st.evm.EpochNumber) // s3 includes homestead
+
+	// Pay intrinsic gas
+	gas, err := IntrinsicGas(st.data, false, homestead)
+	if err != nil {
+		return 0, err
+	}
+	if err = st.useGas(gas); err != nil {
+		return 0, err
+	}
+
+	// Increment the nonce for the next transaction
+	st.state.SetNonce(msg.From(), st.state.GetNonce(sender.Address())+1)
+	st.refundGas()
```

Comment:
If any gas needs to be refunded, it will only be known after the actual operation of the txn is done. So refundGas should happen after the staking operations.

Reply:
Moved. But just want to emphasize that there is no different effect of moving position.